pyriemann/embedding.py

```diff
@@ -92,13 +111,293 @@ def fit_transform(self, X, y=None):
+    return B
+
+
+def locally_linear_embedding(X,
+                             *,
+                             n_components=2,
+                             n_neighbors=5,
+                             metric='riemann',
+                             reg=1e-3):
+    """Riemannian Locally Linear Embedding (LLE).
+
+    Riemannian LLE as proposed in [1]_. LLE is a non-linear, neighborhood-
+    preserving dimensionality reduction algorithm which consists of three
+    main steps:
+    For each point x,
+    1.  find its k nearest neighbors KNN(x) and
+    2.  calculate the best reconstruction of x based on its KNN.
+    3.  Then calculate a low-dimensional embedding for all points based on
+        the weights in step 2.
+
+    Parameters
+    ----------
+    X : ndarray, shape (n_matrices, n_channels, n_channels)
+        Set of SPD matrices.
+    n_components : int, default: 2
+        Dimensionality of projected space.
+    n_neighbors : int, default: 5
+        Number of neighbors for reconstruction of each point.
+    metric : {'riemann', 'logeuclid', 'euclid'}, default: 'riemann'
+        Metric used for KNN and Kernel estimation.
+    reg : float, default: 1e-3
+        Regularization parameter.
+
+    Returns
+    -------
+    embd_ : ndarray, shape (n_matrices, n_components)
+        Locally linear embedding of matrices in X.
+    error_ : float
+        Error of the projected embedding.
+
+    Notes
+    -----
+    .. versionadded:: 0.2.8
+    See sklearn.manifold._locally_linear.locally_linear_embedding for original
+    code.
+
+    References
+    ----------
+    .. [1] A. Goh and R. Vidal, "Clustering and dimensionality reduction
+        on Riemannian manifolds", 2008 IEEE Conference on Computer Vision
+        and Pattern Recognition, June 2008
+    """
+    n_matrices, n_channels, n_channels = X.shape
+    pairwise_distances = pairwise_distance(X, metric=metric)
+    neighbors = np.array([np.argsort(dist)[1:n_neighbors + 1]
+                          for dist in pairwise_distances])
+
+    B = barycenter_weights(X, X, neighbors, metric=metric, reg=reg)
+
+    indptr = np.arange(0, n_matrices * n_neighbors + 1, n_neighbors)
+    W = csr_matrix((B.ravel(), neighbors.ravel(), indptr), shape=(n_matrices,
+                                                                  n_matrices))
+    # M = (W - I).T * (W - I) = W.T * W - W.T - W + I
+    # calculated in the two following lines
+    M = (W.T * W - W.T - W).toarray()
+    M.flat[:: M.shape[0] + 1] += 1
+
+    eigen_values, eigen_vectors = eigh(
+        M, eigvals=(1, n_components), overwrite_a=True
+    )
+    index = np.argsort(np.abs(eigen_values))
+    embd_, error_ = eigen_vectors[:, index], np.sum(eigen_values)
+
+    return embd_, error_
+
+
+def _check_dimensions(X, Y=None, n_components=None, n_neighbors=None):
+    n_matrices, n_channels, n_channels = X.shape
+
+    if not isinstance(Y, type(None)):
+        assert Y.shape[1:] == (n_channels, n_channels), "Dimension of " \
```

Comment:
A general remark: in your code, can you complete error messages, with the numerical values?

Example:
`Y does not have the good number of channels. Should be %d but got %d.' % (n_channels, Y.shape[1])`